htsohm/htsohm.py

```diff
@@ -132,6 +132,83 @@ def select_parent(run_id, max_generation, generation_limit):
+    """"""
+    simulations = config['material_properties']
+    queries = []
+    if 'gas_adsorption_0' in simulations or 'gas_adsorption_1' in simulations:
+        queries.append( getattr(Material, 'gas_adsorption_bin') )
+    if 'surface_area' in simulations:
+        queries.append( getattr(Material, 'surface_area_bin') )
+    if 'helium_void_fraction' in simulations:
+        queries.append( getattr(Material, 'void_fraction_bin') )
+
+    # Each bin is counted...
+    bins_and_counts = session \
+        .query(
+            func.count(Material.id),
+            *queries
+        ) \
+        .filter(
+            Material.run_id == run_id,
+            or_(Material.retest_passed == True, Material.retest_passed == None),
+            Material.generation <= max_generation,
+            Material.generation_index < generation_limit,
+        ) \
+        .group_by(*queries).all()[1:]
+
+    print('\n  BIN\t\t|  COUNT')
+    print('----------------+-----------')
+    for bin_and_count in bins_and_counts:
+        print('  {}\t|  {}'.format(bin_and_count[1:len(queries) + 2], bin_and_count[0]))
+    print('\n')
+
+    input('Press Enter to calculate normalized weights...')
+
+    bins = []
+    for i in bins_and_counts:
+        some_bin = {}
+        for j in range(len(queries)):
+            some_bin[queries[j]] = i[j + 1]
+        bins.append(some_bin)
+    total = sum([i[0] for i in bins_and_counts])
+    # ...then assigned a weight.
+    weights = [ total / float(i[0]) for i in bins_and_counts ]
+    normalized_weights = [ weight / sum(weights) for weight in weights ]
+
+    print('\n  BIN\t\t|  COUNT\t|  WEIGHT\t|  NORMALIZED-WEIGHT')
+    print('----------------+---------------+---------------+---------------------')
+    for i in range(len(bins_and_counts)):
+        print('  {}\t|  {}\t\t|  {}\t\t|  {}'.format(
+            bins_and_counts[i][1:len(queries) + 1], bins_and_counts[i][0],
+            weights[i], normalized_weights[i]))
+    print('\n')
+
+    ga_bin = input('Press select a gas adsorption bin :\t')
+    sa_bin = input('Press select a surface area bin :\t')
+    vf_bin = input('Press select a void fraction bin :\t')
+
+    parent_query = session \
+        .query(Material.uuid) \
+        .filter(
+            Material.run_id == run_id,
+            or_(Material.retest_passed == True, Material.retest_passed == None),
+            Material.gas_adsorption_bin == ga_bin,
+            Material.surface_area_bin == sa_bin,
+            Material.void_fraction_bin == vf_bin,
+            Material.generation <= max_generation,
+            Material.generation_index < generation_limit,
+        ).all()
+    potential_parents = [i[0] for i in parent_query]
+
+    print('Potential parents :')
+    for i in potential_parents:
+        print('\t{}'.format(i))
+
+    parent_uuid = input('Please select a parent UUID :\t')
+
+    return session.query(Material.id).filter(Material.run_id == run_id, Material.uuid == parent_uuid).one()[0]
+
```

Comment:
All this code seems mostly duplicated from select_parent. We have to find a way for this code to be DRY, or it will be a problem going forward